### PR TITLE
Sync Main branch with WebSockets branch.

### PR DESCRIPTION
**Pull request by LarryOsterman**
# Merge changes from Main branch to the feature/websockets branch.

Unlike previous PRs, there *are* significant changes to the WInHttp transport to reflect the different use patterns for WebSockets.

Places to look for changes:

azure\core\azure-core\src\azure\core\http
azure\core\azure-core\src\environment_log_level_listener.cpp (changed code to clean up logging in multithreaded environments)
azure\core\azure-core\inc\azure\core\http

It's unfortunate but the changes made to win_http_transport.cpp include both the existing changes in main and the accommodations needed to handle websockets :(.

## Pull Request Checklist

Please leverage this checklist as a reminder to address commonly occurring feedback when submitting a pull request to make sure your PR can be reviewed quickly:

See the detailed list in the [contributing guide](https://github.com/Azure/azure-sdk-for-cpp/blob/main/CONTRIBUTING.md#pull-requests).

- [ ] [C++ Guidelines](https://azure.github.io/azure-sdk/cpp_introduction.html)
- [ ] Doxygen docs
- [ ] Unit tests
- [ ] No unwanted commits/changes
- [ ] Descriptive title/description
  - [ ] PR is single purpose
  - [ ] Related issue listed
- [ ] Comments in source
- [ ] No typos
- [ ] Update changelog
- [ ] Not work-in-progress
- [ ] External references or docs updated
- [ ] Self review of PR done
- [ ] Any breaking changes?

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/http/win_http_transport.hpp [hunk 1]
```diff
@@ -6,7 +6,6 @@
  * @brief #Azure::Core::Http::HttpTransport implementation via WinHTTP.
  * cspell:words HCERTIFICATECHAIN PCCERT CCERT HCERTCHAINENGINE HCERTSTORE
```
Comment (antkmsft) [inline sdk/core/azure-core/inc/azure/core/http/win_http_transport.hpp:7]: cspell comments should not be in the doxygen comment sections. It was almost correct before (cspell directive was as a regular comment below, but doxygen section also had it).

**Review by antkmsft — APPROVED** [review]
It is hard to review these many changes. If there anything specific that needs review, LMK.
Is my understanding that you pulled changes from main and merging them into a feature branch, and it does not have unique changes correct? But then I see no Identity changes that went in recently...

**Comment by LarryOsterman** [comment]
> It is hard to review these many changes. If there anything specific that needs review, LMK. Is my understanding that you pulled changes from main and merging them into a feature branch, and it does not have unique changes correct? But then I see no Identity changes that went in recently...

See the pull request comment. I recently updated it to include exactly what was changed. Basically there were huge changes to the WinHTTP transport to accommodate the fact that WebSockets can have simultaneous read, write and close operations simultaneously while HTTP can only have a single operation at a time.

**Comment by antkmsft** [comment]
@LarryOsterman, ah I see, thank you!

I compared changes from your personal branch to main, LGTM.

**Pull request merged**